README
Astro::SLA -- interface to C or Fortran SLALIB library
==========

This module provides an interface layer to the C or Fortran version of
the SLALIB library.  From the abstract of the SLALIB manual (written
by Pat Wallace):

   "SLALIB is a library used by writers of positional-astronomy
   applications. Most of the 171 routines are concerned with
   astronomical position and time, but a number have wider
   trigonometrical, numerical or general applications."

An example program, stime, is provided. This simply reports the
current time (modified Julian day and local sidereal time for the JCMT).

Requirements:
-------------

This module requires either the C or Fortran version of the SLALIB
library written by Pat Wallace.

This module has been tested with the Starlink Fortran library v2.5
(released 2005), the current Starlink release (hawaiki),
and 2008 releases of the C library. It should work
fine with older versions (up to 2002).

If you are working with orbital elements you need a version of the
library released sometime in 2002. Recent improvements (mid-2005) to
the accuracy of the calculations for moving sources (comets and
asteroids) means that not all the tests will pass for the orbital
elements tests if you are using a relatively old Slalib.

You must have either the C version of the library or the Starlink
Fortran version of the library in order to build this module.

The Fortran version of SLALIB is available from Starlink under the
Starlink Software Licence (effectively meaning free for non-commercial
use). You can download it from Starlink (http://www.starlink.rl.ac.uk)
using the Starlink Software Store
(http://www.starlink.rl.ac.uk/Software/software_store.htm).
Specifically: http://www.starlink.rl.ac.uk/cgi-store/storeform1?SLA
The GPL Fortran version of slalib is available from the Starlink developer
site (http://dev.starlink.ac.uk).

Please contact Pat Wallace (ptw@tpsoft.demon.co.uk) for details of how
to obtain the (proprietary) C SLALIB library.

A C compiler is required - this is an XS module. A Fortran compiler (such
as g77) is required if you are building against the Fortran library [or
at least the Fortran runtime libraries].

This module has been tested on linux and Mac OS X but should be fairly
portable.


Installation:
-------------

 - Edit the Build.PL so that the location of the SLA
   library and include files are correct. [see variables $sla_lib and
   $sla_inc]. By default the module attempts to find the Starlink
   installation and then falls back to assuming the C version is
   installed. Change the value of $use_fortran if you know you are
   using an explicit C or Fortran version.

   It is assumed the C library is available from /usr/local/lib by default.

   The includes files, slalib.h, should be available
   in the default path. If they are somewhere else the Build.PL
   should be edited to reflect this location. For example,
   change the line

     my $sla_inc => '',

   to

     my $sla_inc => '-I/home/me/include',

   to pick them up from /home/me/include

Then:

  % perl Build.PL
  % ./Build
  % ./Build test
  % ./Build install

The tests only test part of the build. Some functions remain untested.
In particular orbital element support will not be very good unless you
have a modern installation of SLALIB (sometime from 2002).

See Astro::Coords for a general object-oriented wrapper around Astro::SLA.

Documentation:
--------------

The module provides documentation on how to use the perl interface
to SLALIB. It does not contain information on how to use
SLALIB itself. For more information on SLALIB see:

  http://www.starlink.ac.uk/star/docs/sun67.htx/sun67.html


Copyright
---------

Copyright (C) 2007, 2010 Tim Jenness and the Science and Technology Facilities
Council. Copyright (C) 1998-2005 Tim Jenness and the Particle Physics and
Astronomy Research Council.  All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful,but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place,Suite 330, Boston, MA  02111-1307, USA


The SLALIB library (C version) is proprietary.  Please contact Patrick
Wallace (ptw@tpsoft.demon.co.uk) if you would like to obtain a copy.

The Starlink SLALIB library (Fortran) is available from Starlink
under the GPL.

The array handling code was written by Karl Glazebrook.


Changes
-------

See the ChangeLog file for a full list of changes.

Changes for release 1.00:

  - Use Module::Build

  - Fix slaIntin when using the Fortran library

  - Minor compiler warnings fixed and removedsome OUTPUT sections that
    were not needed.

Changes for release 0.99:

  - slaObs can now be called with a constant string (required by
    Astro::Telescope).

  - Added slaIntin (for completeness)

  - The tests have been updated considerably. All the examples
    from the slalib documentation are now included in the test suite.

  - Test::More is now required for the tests.

  - stime has added arguments (-epoch, -h and -man)

Changes for release 0.98:

  - Missed two "Perl_croak + threading errors"

Changes for release 0.97:

  - Addition of a test for cometary ephemerides

  - A fix to the slaOap function when using the Fortran bindings

  - Fix for building with a thread-enabled perl, although slalib
    is not proven to be thread safe.

Changes for release 0.96:

  - Now includes Fortran support

  - Add slaPertel and slaPertue

  - Fix slaOap

Changes for release 0.95:

  - Fix bug in slaPlante (it simply didn't work)

Changes for release 0.94:

  - Fix bug in slaPreces and add new test for it.

Changes for release (0.93):

  - New utility 'stime' for printing the current time, MJD and LST

  - Add ut2lst_tel command (this was in the docs but not present)

  - Fix bug in slaEg50 (thanks to Chris Phillips)

  - Fix segmentation violation in certain uses of the slaObs
    command on linux.

  - Should now compile cleanly on perl 5.6.0 (now uses PL_na symbols
    in XS layer)

  - slightly improved test suite.

Changes for release 0.92:

  - Export all single precision aliases

  - Export slaDvxv

  - Fix compiler warnings (thanks to Chris Phillips <phillips@jive.nfra.nl>
    for a patch).

Changes for release 0.91:

  - Add DR2S, DR2H and slaEcleq to the export list.

  - Fix argument order of slaEqgal and slaSupgal
        [patch supplied by Tim Kimball (kimball@stsci.edu)]

  - Allow use of single precision names as aliases to double precision
    commands.


ChangeLog
2010-12-30  Tim Jenness  <timj@poma.jach.hawaii.edu>

	* Build.PL: Use Module::Build

2007-07-24  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Fix slaIntin issues for Fortran vs C interface. Tidy up
	slaIntin test.

	* SLA.xs: Remove some unnecessary OUTPUT sections.

2005-09-22  Tim Jenness  <tjenness@cpan.org>

	---- V0.99 CPAN ---

2005-09-02  Tim Jenness  <tjenness@cpan.org>

	* t/mjd.t: Switch to Test::More

	* t/preces.t: Switch to Test::More

	* t/slaObs.t: Switch to Test::More

	* t/fk4to5.t: Switch to Test::More

	* t/comet.t: Add 3200 Phaethon torture test and switch to Test::More

2004-02-03  Tim Jenness  <tjenness@cpan.org>

	* stime: Tidy up. Add -epoch, -h and -man options.

	* SLA.pm (ut2lst): Tidy up.

2004-02-01  Tim Jenness  <tjenness@cpan.org>

	* t/sun67.t: Add 70 new tests based on the examples in SUN/67

2004-01-31  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Mods for slaObs update. Added support for slaIntin

	* SLA.pm (slaObs): Can now be called with a constant string as
	second argument. [also modified slaObs.t to match]

2003-10-22  Tim Jenness  <tjenness@cpan.org>

	--- VERSION 0.98 on CPAN ---

2003-09-04  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Missed two Perl_croaks that only occur when linking
	against the C version of SLA.

2003-08-18  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Release Version 0.97 on CPAN

2003-07-14  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Need to add aTHX_ for Perl_croak to work with threads

2003-07-03  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Fix error in slaOap fortran interface. This was breaking
	the comet test. Does not affect the C version.

2003-07-02  Tim Jenness  <tjenness@cpan.org>

	* t/comet.t: Add a test for perturbation of orbital elements since
	that is key functionality for some users.

2002-09-15  Tim Jenness  <tjenness@cpan.org>

	--- VERSION 0.96 on CPAN ---

	* SLA.pm: Announce Fortran version support in documentation

2002-09-15  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Add full support for Fortran version of SLALIB

2002-08-15  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Fix argument list for slaOap


2002-08-11  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Add slaPertel and slaPertue

2002-05-24  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Make release 0.95

2002-04-12  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: slaPlante had wrong calling arguments

2001-04-24  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: update copyright

	* MANIFEST: Add t/preces.t

	* README: Tweak for 0.94

	* SLA.pm: Version 0.94. Update copyright.

2000-11-21  Tim Jenness  <tjenness@cpan.org>

	* t/preces.t: Add test for slaPreces

	* SLA.xs: slaPreces - the ra and dec variables are input and
	output not just output.

	* t/fk4to5.t: Use Test module

2000-04-11  Tim Jenness  <tjenness@cpan.org>

	* t/fk4to5.t: Improve the test (and fix btoj for -00 decs)
	(btoj): dont use abs($d), multiply by dsign instead.

2000-03-28  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm (ut2lst_tel): Add missing command

	* t/mjd.t: Add new test

	* Makefile.PL: Add Test prerequisite

	* SLA.pm: Increase version number to 0.93

	* t/slaObs.t: Add new test.

	* SLA.xs: Rename slaObs to _slaObs and call from perl layer

	* SLA.pm (slaObs): Add perl layer to translate undef to '' to
	prevent segmentation violation occurring on my linux system.

	* stime: Add example program

        * SLA.pm (lstnow_tel): Trap for unknown telescopes.

2000-03-24  Tim Jenness  <tjenness@cpan.org>

	* arrays.c: Use PL_na symbols rather than na

2000-02-14  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs: Apply patch to slaEg50 thanks to Chris Phillips.

1999-10-11  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Version number now 0.92

	* SLA.pm: Export single precisions commands. Export slaDvxv.
	Update documentation to reflect the exportable commands rather
	than the initial list.

	* SLA.xs: Fix compiler warnings by using void* (patch supplied by
	Chris Phillips <phillips@jive.nfra.nl>)

1999-08-03  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Version now 0.91

	* SLA.xs: Apply patch for slaEqgal and slaSupgal supplied by
	Tim Kimball (kimball@stsci.edu)
	Allow use of single precision names as aliases to double precision
	commands.
	Remove use of static char.

	* SLA.pm: Add DR2S and DR2H to the export list.

1999-05-17  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Version now 0.90 in preparation for CPAN release.

1999-03-12  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Up version number to 0.11

	* SLA.xs: slaDmoon fixed - double not float

1998-06-02  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Complete 99% + add lst routines

	* SLA.xs: Complete 99%


1998-05-27  Tim Jenness  <tjenness@cpan.org>

	* SLA.xs (slaDd2tf): Add slaDd2tf and slaDjcl

1998-04-01  Tim Jenness  <tjenness@cpan.org>

	* SLA.pm: Added slaPreces

	* SLA.xs (slaPreces): Added slaPreces



Build.PL
use strict;
use warnings;
use File::Spec;
use Module::Build;

# A makefile can be generated for either the C or Fortran version of
# SLALIB. It does not yet determine the C/Fortran-ness of your library
# automatically.

# Change this flag to control the behaviour of the build.
# This controls whether we are using the Fortran version (1),
# the C version (0) or the Starlink Fortran version (-1)
my $use_fortran = -1;

# Default location for the library. This is used if we fail to find
# starlink fortran (in which case it is assumed to be a C library)
# or use_fortran is set to 0 or 1
# If you have an explicit non-standard path prepend -L/wherever
my $sla_lib = ' -L/usr/local/lib -lsla ';

# Same for include files [only relevant for C version] [use -I]
my $sla_inc = '-I/usr/local/include';

# Check valid range for the fortran switch
if ($use_fortran != 0 && $use_fortran != -1 && $use_fortran != 1) {
  print "use_fortran variable set to strange value. Assuming -1 [Starlink]\n";
}

# Starlink
if ($use_fortran == -1) {
  use vars qw/ %StarConfig /;

  # See if we can find a starlink system
  # $STARLINK env overrides everything
  if (exists $ENV{STARLINK} && defined $ENV{STARLINK}) {
    print "Using STARLINK environment variable: $ENV{STARLINK}\n";
    %StarConfig = (
		   Star_Inc => $ENV{STARLINK}."/include",
		   Star_Lib => $ENV{STARLINK}."/lib",
		   Star_Bin => $ENV{STARLINK}."/bin",
                  );
  } else {
    # Must look for it
    eval "use Starlink::Config";
    if ($@) {
      # do not have a Starlink configuration so guess /star
      %StarConfig = (
  		     Star_Inc => '/star/include',
		     Star_Lib => '/star/lib',
		     Star_Bin => '/star/bin',
		    );
    } else {
      print "Found Starlink configuration. Using $StarConfig{Star}\n";
    }
  }

  # Work out the library full path
  my $lib = File::Spec->catfile($StarConfig{Star_Lib}, "libsla.a" );

  if (-e $lib) {
    print "Located Starlink SLALIB library in $StarConfig{Star_Lib}\n";
    $sla_lib = " -L$StarConfig{Star_Lib} -lsla ";

    # See if we can find the version
    eval { require Starlink::Versions; };
    if (!$@) {
      my $ver = Starlink::Versions::starversion_string('sla');
      $ver = ( defined $ver ? $ver : "****** undefined ******" );
      print "It seems to be SLA version $ver\n";
    }

  } else {
    # Assume the C fallback
    $use_fortran = 0;
  }


}



# The #defines, libraries and any additional prerequisites
my (@defines, $libs, %build_prereqs);
if ($use_fortran) {
  print "Attempting to build against a Fortran SLALIB: $sla_lib\n";

  # Must switch modes
  push(@defines, "-DUSE_FORTRAN");

  # Must include fortran libraries but we do not want to force
  # ExtUtils::F77 on an unsuspecting C user
  $build_prereqs{'ExtUtils::F77'} = 0;

  # use eval here since we are listing ExtUtils as a prerequisite
  # anyway and so if this does not work we'll still get a message
  # later
  my $flibs = '';
  eval { require ExtUtils::F77; ExtUtils::F77->import };
  if ($@) {
    print "Error determining Fortran library requirements\n";
  } else {
    $flibs = ExtUtils::F77->runtime;
    push(@defines, "-DHAS_UNDERSCORE")
      if ExtUtils::F77->trail_;
  }

  $libs = "$sla_lib $flibs";

} else {
  print "Attempting to build against a C SLALIB library: $sla_lib\n";

  # no special defines for the C version

  # Just need the math library
  $libs = "$sla_lib -lm";
}

my $file = File::Spec->catfile( "lib", "Astro", "SLA.pm" );

my $build = Module::Build->new
  (
   module_name => "Astro::SLA",
   abstract_from => $file,
   license => "gpl",
   author => [
              'Tim Jenness <tjenness@cpan.org>',
             ],
   version_from => $file,
   meta_merge => {
               resources =>  {
                              repository => "git://github.com/timj/perl-Astro-SLA.git",
                              homepage => "http://github.com/timj/perl-Astro-SLA/tree/master",
                             },
                },
   script_files => [ qw/
                         stime
                       /],
   requires => {
                "Pod::Usage" => 0,
                "Getopt::Long" => 0,
               },
   build_requires => {
                      "Test::More" => 0,
                      %build_prereqs,
                     },
   c_source => [ "src" ],
   extra_compiler_flags => [$sla_inc, @defines],
   extra_linker_flags => $libs,
  );

$build->create_build_script;



lib/Astro/SLA.pm
package Astro::SLA;

=head1 NAME

Astro::SLA - Perl interface to SLAlib positional astronomy library

=head1 SYNOPSIS

  use SLA;
  use SLA qw(:constants :sla);

  slaFk45z($ra, $dec, 1950.0, $ra2000, $dec2000);
  slaCldj($yy, $mn, $dd, $mjd, $status);

  ($lst, $mjd) = lstnow($long);
  ($lst, $mjd) = ut2lst_tel($yy,$mn,$dd,$hh,$mm,$ss,'JCMT');

=head1 DESCRIPTION

This modules provides a Perl interface to either the C or Fortran
versions of the SLALIB astrometry library written by Pat Wallace.

In general the single precision routines have not been
implemented since perl can work in double precision.

The SLALIB constants (as provided in slamac.h) are available.

In addition small utility subroutines are provided that
do useful tasks (from the author's point of view) - specifically
routines for calculating the Local Sidereal Time.

=cut

# '  -- close quote for my 'authors' apostrophe above.

use strict;
use Carp;
use vars qw($VERSION %EXPORT_TAGS);

use Exporter 'import';
use base qw/ DynaLoader /;

$VERSION = '1.00';

%EXPORT_TAGS = (
                'sla'=>[qw/
                            slaAddet slaAfin slaAirmas slaAmp slaAmpqk
                            slaAop slaAoppa slaAoppat slaAopqk slaAtmdsp
                            slaAv2m slaBear slaCaf2r slaCaldj slaCalyd
                            slaCc2s slaCc62s slaCd2tf slaCldj slaClyd
                            slaCr2af slaCr2tf slaCs2c6 slaDaf2r slaDafin
                            slaDat slaDav2m slaDbear slaDbjin slaDc62s
                            slaDcc2s slaDcmpf slaDcs2c slaDd2tf slaDe2h
                            slaDeuler slaDfltin slaDh2e slaDimxv slaDjcal
                            slaDjcl slaDm2av slaDmoon slaDmxm slaDmxv slaDpav
                            slaDr2af slaDr2tf slaDrange slaDranrm slaDs2c6
                            slaDs2tp slaDsep slaDtf2d slaDtf2r slaDtp2s
                            slaDtp2v slaDtps2c slaDtpv2c slaDtt slaDv2tp
                            slaDvdv slaDvn slaDvxv slaE2h slaEarth slaEcleq
                            slaEcmat slaEcor slaEg50 slaEpb slaEpb2d slaEpco
                            slaEpj slaEpj2d slaEqecl slaEqeqx slaEqgal
                            slaEtrms slaEuler slaEvp slaFk425 slaFk45z
                            slaFk54z slaFloatin slaGaleq slaGalsup slaGe50
                            slaGeoc slaGmst slaGmsta slaGresid slaH2e slaImxv
                            slaInvf slaKbj slaM2av slaMap slaMappa slaMapqk
                            slaMapqkz slaMoon slaMxm slaMxv slaNut slaNutc
                            slaOap slaOapqk slaObs slaPa slaPav slaPcd
                            slaPertue slaPertel
                            slaPda2h slaPdq2h slaPlanel slaPlanet slaPlante
                            slaPm slaPolmo slaPrebn slaPrec slaPreces
                            slaPrecl slaPrenut slaPvobs slaRandom slaRange
                            slaRanorm slaRcc slaRdplan slaRefco slaRefcoq
                            slaRefv slaRefz slaRverot slaRvgalc slaRvlg
                            slaRvlsrd slaRvlsrk slaS2tp slaSep slaSubet
                            slaSupgal slaTp2s slaTp2v slaTps2c slaTpv2c
                            slaUnpcd slaV2tp slaVdv slaVxv slaWait slaXy2xy
                            slaZd slaIntin
                          /],

                'constants'=>[qw/
                                  DPI D2PI D1B2PI D4PI D1B4PI DPISQ DSQRPI DPIBY2
                                  DD2R DR2D DAS2R DR2AS DH2R DR2H DS2R DR2S D15B2P
                                /],

                'funcs'=>[qw/
                              lstnow lstnow_tel ut2lst ut2lst_tel
                            /]
               );


Exporter::export_tags('sla','constants','funcs');

bootstrap Astro::SLA;


=head1 Routines

There are 3 distinct groups of routines that can be imported into
the namespace via tags:

=over 4

=item sla - import just the SLALIB routines

=item constants - import the SLALIB constants

=item funcs - import the extra routines

=back

Each group will be discussed in turn.

=head2 sla

All the double precision SLA routines are implemented except for
slaPxy, slaDmat, slaSvd, slaSvdcov, slaSvdsol (I may do these some
other time -- although they should be done in C<PDL>).

The implemented routines are:

 slaAddet slaAfin slaAirmas slaAmp slaAmpqk slaAop
 slaAoppa slaAoppat slaAopqk slaAtmdsp slaAv2m slaBear
 slaCaf2r slaCaldj slaCalyd slaCc2s slaCc62s slaCd2tf
 slaCldj slaClyd slaCr2af slaCr2tf slaCs2c6 slaDaf2r
 slaDafin slaDat slaDav2m slaDbear slaDbjin slaDc62s
 slaDcc2s slaDcmpf slaDcs2c slaDd2tf slaDe2h slaDeuler
 slaDfltin slaDh2e slaDimxv slaDjcal slaDjcl slaDm2av
 slaDmoon slaDmxm slaDmxv slaDpav slaDr2af slaDr2tf
 slaDrange slaDranrm slaDs2c6 slaDs2tp slaDsep slaDtf2d
 slaDtf2r slaDtp2s slaDtp2v slaDtps2c slaDtpv2c slaDtt
 slaDv2tp slaDvdv slaDvn slaDvxv slaE2h slaEarth slaEcleq
 slaEcmat slaEcor slaEg50 slaEpb slaEpb2d slaEpco slaEpj
 slaEpj2d slaEqecl slaEqeqx slaEqgal slaEtrms slaEuler
 slaEvp slaFk425 slaFk45z slaFk54z slaFloatin slaGaleq
 slaGalsup slaGe50 slaGeoc slaGmst slaGmsta slaGresid slaH2e
 slaImxv slaInvf slaKbj slaM2av slaMap slaMappa slaMapqk
 slaMapqkz slaMoon slaMxm slaMxv slaNut slaNutc slaOap
 slaOapqk slaObs slaPa slaPav slaPcd slaPda2h slaPdq2h
 slaPlanel slaPlanet slaPlante slaPm slaPolmo slaPrebn
 slaPrec slaPreces slaPrecl slaPrenut slaPvobs slaRandom
 slaRange slaRanorm slaRcc slaRdplan slaRefco slaRefcoq
 slaRefv slaRefz slaRverot slaRvgalc slaRvlg slaRvlsrd
 slaRvlsrk slaS2tp slaSep slaSubet slaSupgal slaTp2s slaTp2v
 slaTps2c slaTpv2c slaUnpcd slaV2tp slaVdv slaVxv slaWait
 slaXy2xy slaZd slaIntin

Also, slaGresid and slaRandom are not in the C library (although they
are in the Fortran version).  slaWait is implemented using the perl
'select(ready file descriptors)' command.

In general single precision routines are simply aliases of the
double precision equivalents.

For more information on the SLALIB routines consult the Starlink
documentation (Starlink User Note 67 (SUN/67)). This document
is available from the Starlink web site (http://www.starlink.ac.uk/)
[SUN67 available from:
http://www.starlink.ac.uk/cgi-bin/htxserver/sun67.htx/sun67.html ]


=cut

# Implement at the perl level (command is unimplemented at C level)

sub slaWait ($) {
  my $delay = shift;
  select (undef, undef, undef, $delay);

}


# slaObs
#   In order to overcome a segmentation violation under linux (at least)
#   occuring when 'c' is set to undef but is to be a return value
#   have a perl layer that replaces undef with '' in order to fix the
#   issue
# We also do this so that a constant can be supplied as the argument.
# undef first argument is converted to -1

sub slaObs ($$$$$$) {
  my $c = (defined $_[1] ? $_[1] : '');
  my $outc;
  my $n = (defined $_[0] ? $_[0] : -1);
  _slaObs($n, $c, $outc, $_[2], $_[3], $_[4], $_[5]);

  # Copy outc to the caller namespace only if a positive
  # number was specified to slaObs
  if ($n > 0) {
    $_[1] = $outc;
  }

  return;
}

=head2 Constants

Constants supplied by this module (note that they are implemented via the
L<constant> pragma):

=over 4

=item DPI - Pi

=item D2PI - 2 * Pi

=item D1B2PI - 1 / (2 * Pi)

=item D4PI - 4 * Pi

=item D1B4PI - 1 / (4 * Pi)

=item DPISQ - Pi ** 2 (Pi squared)

=item DSQRPI - sqrt(Pi)

=item DPIBY2 - Pi / 2: 90 degrees in radians

=item DD2R - Pi / 180: degrees to radians

=item DR2D - 180/Pi:  radians to degrees

=item DAS2R - pi/(180*3600): arcseconds to radians

=item DR2AS - 180*3600/pi: radians to arcseconds

=item DH2R - pi/12: hours to radians

=item DR2H - 12/pi: radians to hours

=item DS2R - pi / (12*3600): seconds of time to radians

=item DR2S - 12*3600/pi: radians to seconds of time

=item D15B2P - 15/(2*pi): hours to degrees * radians to turns

=back

=cut


# Could implement these directly via the include file in the XS layer.
# Since these cant change - implement them explicitly.

# Pi
use constant DPI => 3.1415926535897932384626433832795028841971693993751;

# 2pi
use constant D2PI => 6.2831853071795864769252867665590057683943387987502;

# 1/(2pi)
use constant D1B2PI => 0.15915494309189533576888376337251436203445964574046;

# 4pi
use constant D4PI => 12.566370614359172953850573533118011536788677597500;

# 1/(4pi)
use constant D1B4PI => 0.079577471545947667884441881686257181017229822870228;

# pi^2
use constant DPISQ => 9.8696044010893586188344909998761511353136994072408;

# sqrt(pi)
use constant DSQRPI => 1.7724538509055160272981674833411451827975494561224;

# pi/2:  90 degrees in radians
use constant DPIBY2 => 1.5707963267948966192313216916397514420985846996876;

# pi/180:  degrees to radians
use constant DD2R => 0.017453292519943295769236907684886127134428718885417;

# 180/pi:  radians to degrees
use constant DR2D => 57.295779513082320876798154814105170332405472466564;

# pi/(180*3600):  arcseconds to radians
use constant DAS2R => 4.8481368110953599358991410235794797595635330237270e-6;

# 180*3600/pi :  radians to arcseconds
use constant DR2AS => 2.0626480624709635515647335733077861319665970087963e5;

# pi/12:  hours to radians
use constant DH2R => 0.26179938779914943653855361527329190701643078328126;

# 12/pi:  radians to hours
use constant DR2H => 3.8197186342054880584532103209403446888270314977709;

# pi/(12*3600):  seconds of time to radians
use constant DS2R => 7.2722052166430399038487115353692196393452995355905e-5;

# 12*3600/pi:  radians to seconds of time
use constant DR2S => 1.3750987083139757010431557155385240879777313391975e4;

# 15/(2pi):  hours to degrees x radians to turns
use constant D15B2P => 2.3873241463784300365332564505877154305168946861068;


=head2 Extra functions (using the 'funcs' tag)


=over 4

=item B<lstnow_tel>

Return current LST (in radians) and MJD for a given telescope.
The telescope identifiers should match those present in slaObs.
The supplied telescope name is converted to upper case.

   ($lst, $mjd) = lstnow_tel($tel);

Aborts if telescope name is unknown.

=cut

sub lstnow_tel {

  croak 'Usage: lstnow_tel($tel)' unless scalar(@_) == 1;

  my $tel = shift;

  # Upper case the telescope
  $tel = uc($tel);

  my ($w, $p, $h, $name);

  # Find the longitude of this telescope
  slaObs(-1, $tel, $name, $w, $p, $h);

  # Check telescope name
  if ($name eq '?') {
    croak "Telescope name $tel unrecognised by slaObs()";
  }

  # Convert longitude to west negative
  $w *= -1.0;

  # Run lstnow
  lstnow($w);

}


=item B<lstnow>

Return current LST (in radians) and MJD (days)
Longitude should be negative if degrees west
and in radians.

  ($lst, $mjd) = lstnow($long);

=cut

sub lstnow {

  croak 'Usage: lstnow($long)' unless scalar(@_) == 1;

  my $long = shift;

  my ($sign, @ihmsf);

  # Get current UT time
  my ($sec, $min, $hour,$mday,$mon,$year,$wday,$yday,$isdst) = gmtime(time);

  # Calculate LST
  $year += 1900;
  $mon++;
  my ($lst, $mjd) = ut2lst($year, $mon, $mday, $hour, $min, $sec, $long);

  return ($lst, $mjd);

}


=item B<ut2lst>

Given the UT time, calculate the Modified Julian date (UTC) and the
local sidereal time (radians) for the specified longitude.

 ($lst, $mjd) = ut2lst(yy, mn, dd, hh, mm, ss, long)

Longitude should be negative if degrees west and in radians.

=cut

sub ut2lst {

  croak 'Usage: ut2lst(yy,mn,dd,hh,mm,ss,long)'
    unless scalar(@_) == 7;

  my ($yy, $mn, $dd, $hh, $mm, $ss, $long) = @_;

  # Calculate fraction of day
  slaDtf2d($hh, $mm, $ss, my $fd, my $j);
  if ($j != 0) {
    croak "Error calculating fractional day with H=$hh M=$mm S=$ss\n";
  }

  # Calculate modified julian date of UT day
  slaCldj($yy, $mn, $dd, my $mjd, my $slastatus);

  if ($slastatus != 0) {
    croak "Error calculating modified Julian date with args: $yy $mn $dd\n";
  }

  # Calculate sidereal time of greenwich
  my $gmst = slaGmsta($mjd, $fd);

  # Find MJD of current time (not just day)
  $mjd += $fd;

  # Equation of the equinoxes (requires TT although makes very
  # little differnece)
  my $tt = $mjd + ( slaDtt($mjd) / 86_400.0);
  my $eqeqx = slaEqeqx($tt);

  # Local sidereal time = GMST + EQEQX + Longitude in radians
  my $lst = slaDranrm($gmst + $eqeqx + $long);

  return ($lst, $mjd);
}

=item B<ut2lst_tel>

Given the UT time, calculate the Modified Julian date and the
local sidereal time (radians) for the specified telescope.

 ($lst, $mjd) = ut2lst_tel(yy, mn, dd, hh, mm, ss, tel)

=cut

sub ut2lst_tel ($$$$$$$) {
  croak 'Usage: ut2lst_tel($tel)' unless scalar(@_) == 7;

  my $tel = pop(@_);

  # Upper case the telescope
  $tel = uc($tel);

  my ($w, $p, $h, $name);

  # Find the longitude of this telescope
  slaObs(-1, $tel, $name, $w, $p, $h);

  # Check telescope name
  if ($name eq '?') {
    croak "Telescope name $tel unrecognised by slaObs()";
  }

  # Convert longitude to west negative
  $w *= -1.0;

  # Run ut2lst
  return ut2lst(@_, $w);

}

=back


=head1 AUTHOR

Tim Jenness E<gt>tjenness@cpan.orgE<lt>

=head1 REQUIREMENTS

The Fortran version of SLALIB is available from Starlink under the
Gnu GPL. You can download it as part of the Starlink software
collection (http://www.starlink.rl.ac.uk).

The SLALIB library (C version) is proprietary.  Please contact Patrick
Wallace (ptw@tpsoft.demon.co.uk) if you would like to obtain a copy.

=head1 COPYRIGHT

Copyright (C) 2007, 2010 Tim Jenness and the Science and Technology Facilities
Council. Copyright (C) 1998-2005 Tim Jenness and the Particle Physics and
Astronomy Research Council.  All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful,but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place,Suite 330, Boston, MA  02111-1307, USA

=cut

1;
